packages/libs/error-stack/src/serde/hook.rs

```diff
@@ -0,0 +1,360 @@
+use alloc::{boxed::Box, vec::Vec};
+use core::{any::TypeId, iter::FusedIterator};
+
+pub(crate) use default::install_builtin_hooks;
+
+use crate::Frame;
+
+pub struct Serde {}
+
+pub type HookContext<T> = crate::hook::context::HookContext<Serde, T>;
+
+fn serialize<'a, T: serde::Serialize + Send + Sync + 'static>(
+    frame: &'a Frame,
+) -> Option<Box<dyn erased_serde::Serialize + 'a>> {
+    #[cfg(nightly)]
+    let value: &T = frame.request_ref()?;
+    #[cfg(not(nightly))]
+    let value: &T = frame.downcast_ref()?;
+    Some(Box::new(value))
+}
+
+// Thanks to https://users.rust-lang.org/t/hrtb-on-multiple-generics/34255/2 for the solution
+pub trait DynamicFn<'a, I>: Send + Sync + 'static
+where
+    I: Send + Sync + 'static,
+{
+    type Output: serde::Serialize + 'a;
+
+    fn call(&self, value: &'a I, context: &mut HookContext<I>) -> Self::Output;
+}
+
+impl<'a, U: 'a, T, F> DynamicFn<'a, T> for F
+where
+    F: Fn(&'a T, &mut HookContext<T>) -> U,
+    F: Send + Sync + 'static,
+    U: serde::Serialize,
+    T: Send + Sync + 'static,
```

Comment:
Is this the only implementation for `DynamicFn`? If yes do you see a way to directly use `Fn` instead without introducing the trait?

Reply:
Yes, `DynamicFn` (I believe I have a comment referring to the post) is needed to specify lifetimes for the return type, which depends on the first argument (using HRTB) for the return type. Otherwise, the return type must be `'static`, which means that custom hooks need to return owned values, unable to borrow from the downcasted/requested value. Without the trait, we're unable to do that ._.